### PR TITLE
DRILL-8155: Introduce New Plugin Authentication Modes

### PR DESCRIPTION
**Pull request by cgivre**
# [DRILL-8155](https://issues.apache.org/jira/browse/DRILL-8155): Introduce new plugin authentication modes

## Description
At present, Drill storage plugins can use a shared set of credentials to access storage on behalf of Drill users or, in a subset of cases belonging to the broader Hadoop family, they can impersonate the Drill user when drill.exec.impersonation.enabled = true.  An important but missing auth mode is what is termed ["user translation" in Trino](https://docs.starburst.io/latest/security/impersonation.html).   Under user translation, the active Drill user is translated to a user known to the external storage by means of a translation table that associates Drill users with their credentials for the external storage.  No support for user impersonation in the external storage is required in this mode.  This ticket proposes that we add establish a design pattern that adds support for this auth mode to Drill storage plugins.

This ticket also adds support for `USER_TRANSLATION` mode for the HTTP and JDBC storage plugins.

Another present day limitation is that impersonation, for the plugins that support it, is toggled by a global switch.  We propose here that the auth mode chosen for a plugin should be independent of the auth modes chosen for other plugins, by a move of this option into their respective storage configs.

Finally, while a standardised means of choosing an authentication mode is desired, note that not every storage plugin needs to, or can, support every mode.

## Documentation
New plugin authentication modes to be documented.

## Testing
Added unit tests.

### CODE REVIEW

File: contrib/storage-jdbc/src/test/java/org/apache/drill/exec/store/jdbc/TestJdbcWriterWithH2.java [hunk 1]
```diff
@@ -437,6 +445,7 @@ public void testWithReallyLongFile() throws Exception {
   }
 
   @Test
+  @Ignore
```
[the remark this answers is not in the document]
Comment (cgivre) [reply contrib/storage-jdbc/src/test/java/org/apache/drill/exec/store/jdbc/TestJdbcWriterWithH2.java:448]: Fixed (and un-ignored)

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcScanBatchCreator.java [hunk 2]
```diff
@@ -55,14 +57,22 @@ public CloseableRecordBatch getBatch(ExecutorFragmentContext context,
...
     ScanFrameworkBuilder builder = new ScanFrameworkBuilder();
     builder.projection(subScan.getColumns());
     builder.setUserName(subScan.getUserName());
     JdbcStoragePlugin plugin = subScan.getPlugin();
-    List<ManagedReader<SchemaNegotiator>> readers =
-      Collections.singletonList(new JdbcBatchReader(plugin.getDataSource(), subScan.getSql(), subScan.getColumns()));
+    UserCredentials userCreds = context.getContextInformation().getQueryUserCredentials();
+    DataSource ds = plugin.getDataSource(userCreds)
+      .orElseThrow(() -> UserException.permissionError().message(
+        "Query user %s could obtain a connection to %s, missing credentials?",
```
[the remark this answers is not in the document]
Comment (cgivre) [reply contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcScanBatchCreator.java:68]: Fixed.

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcConventionFactory.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+package org.apache.drill.exec.store.jdbc;
+
+import org.apache.calcite.sql.SqlDialect;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class JdbcConventionFactory {
+
+  private final Map<SqlDialect, DrillJdbcConvention> CACHE = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (jnturton) [reply contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcConventionFactory.java]: Thanks, I've replaced it with a Guava cache that has a maximum size of 100 before evictions LRU begin. I wonder if I should set an expiry time on it too?

**Review by vvysotskyi — APPROVED** [review]
Thanks for making changes, +1

**Pull request merged**